storage/store/approvals.go

```diff
@@ -43,68 +43,49 @@ func NewResultApprovals(collector module.CacheMetrics, db storage.DB) *ResultApp
 		db: db,
 		cache: newCache(collector, metrics.ResourceResultApprovals,
 			withLimit[flow.Identifier, *flow.ResultApproval](flow.DefaultTransactionExpiry+100),
-			withStore(store),
+			withStoreWithLock(storeWithLock),
 			withRetrieve(retrieve)),
-		indexing: new(sync.Mutex),
 	}
 }
 
 // Store stores a ResultApproval
-func (r *ResultApprovals) Store(approval *flow.ResultApproval) error {
-	return r.db.WithReaderBatchWriter(func(rw storage.ReaderBatchWriter) error {
-		return r.cache.PutTx(rw, approval.ID(), approval)
-	})
-}
-
-// Index indexes a ResultApproval by chunk (ResultID + chunk index).
-// This operation is idempotent (repeated calls with the same value are equivalent to
-// just calling the method once; still the method succeeds on each call).
-//
-// CAUTION: the Flow protocol requires multiple approvals for the same chunk from different verification
-// nodes. In other words, there are multiple different approvals for the same chunk. Therefore, the index
-// Executed Chunk ➜ ResultApproval ID (populated here) is *only safe* to be used by Verification Nodes
-// for tracking their own approvals.
-func (r *ResultApprovals) Index(resultID flow.Identifier, chunkIndex uint64, approvalID flow.Identifier) error {
-	// For the same ExecutionResult, a correct Verifier will always produce the same approval. In other words,
-	// if we have already indexed an approval for the pair (resultID, chunkIndex) we should never overwrite it
-	// with a _different_ approval. We explicitly enforce that here to prevent state corruption.
-	// The lock guarantees that no other thread can concurrently update the index. Thereby confirming that no value
-	// is already stored for the given key (resultID, chunkIndex) and then updating the index (or aborting) is
-	// synchronized into one atomic operation.
-	r.indexing.Lock()
-	defer r.indexing.Unlock()
-
-	err := r.db.WithReaderBatchWriter(func(rw storage.ReaderBatchWriter) error {
-		var storedApprovalID flow.Identifier
-		err := operation.LookupResultApproval(rw.GlobalReader(), resultID, chunkIndex, &storedApprovalID)
-		if err != nil {
-			if !errors.Is(err, storage.ErrNotFound) {
-				return fmt.Errorf("could not lookup result approval ID: %w", err)
-			}
-
-			// no approval found, index the approval
-
-			return operation.UnsafeIndexResultApproval(rw.Writer(), resultID, chunkIndex, approvalID)
-		}
+func (r *ResultApprovals) Store(lctx lockctx.Proof, approval *flow.ResultApproval) error {
```

Comment:
I'd like to the caller to pass the lock context as proof to the Store method, instead of the approvals store holding the lock manager and creating the lock context. 

Why? because it's more flexible, so that the caller can pass the lock context to other modules which might also need synchronization.